src/storage/transaction/TossEdgeIterator.h

```diff
@@ -0,0 +1,227 @@
+ */
+
+#ifndef STORAGE_TRANSACTION_TOSSEDGEITERATOR_H_
+#define STORAGE_TRANSACTION_TOSSEDGEITERATOR_H_
+
+#include <folly/String.h>
+
+#include "common/base/Base.h"
+#include "kvstore/KVIterator.h"
+#include "storage/CommonUtils.h"
+#include "storage/exec/StorageIterator.h"
+#include "storage/transaction/TransactionManager.h"
+#include "storage/StorageFlags.h"
+
+namespace nebula {
+namespace storage {
+
+class TossEdgeIterator : public SingleEdgeIterator {
+public:
+    TossEdgeIterator(
+            PlanContext* planCtx,
+            std::unique_ptr<kvstore::KVIterator> iter,
+            EdgeType edgeType,
+            const std::vector<std::shared_ptr<const meta::NebulaSchemaProvider>>* schemas,
+            const folly::Optional<std::pair<std::string, int64_t>>* ttl,
+            bool moveToValidRecord = true)
+        : SingleEdgeIterator() {
+            planContext_ = planCtx;
+            iter_ = std::move(iter);
+            edgeType_ = edgeType;
+            schemas_ = schemas;
+            if (ttl->hasValue()) {
+                hasTtl_ = true;
+                ttlCol_ = ttl->value().first;
+                ttlDuration_ = ttl->value().second;
+            }
+            moveToValidRecord_ = moveToValidRecord;
+            if (isEdge(iter_->key()) && setReader(iter_->val())) {
+                lastRank_ = NebulaKeyUtils::getRank(planContext_->vIdLen_, iter_->key());
+                lastDstId_ = NebulaKeyUtils::getDstId(planContext_->vIdLen_, iter_->key()).str();
+                LOG_IF(INFO, FLAGS_trace_toss)
+                    << "first edge valid ver="
+                    << NebulaKeyUtils::getVersion(planContext_->vIdLen_, iter_->key());
+            } else {
+                next();
+            }
+        }
+
+    folly::StringPiece key() const override {
+        if (iter_->valid()) {
+            return iter_->key();
+        } else if (recoverEdgesIter_ != recoverEdges_.end()) {
+            auto rLockedPtr = (*recoverEdgesIter_)->rlock();
+            return rLockedPtr->first;
+        }
+        return iter_->key();
+    }
+
+    bool valid() const override {
+        if (!iter_->valid() && recoverEdgesIter_ == recoverEdges_.end()) {
+            LOG_IF(INFO, FLAGS_trace_toss) << "TossEdgeIterator::valid() = false";
+            return false;
+        }
+        auto ret = lookupOne_ && reader_ != nullptr;
+
+        if (FLAGS_trace_toss) {
+            std::string key;
+            std::string val;
+            if (iter_->valid()) {
+                key = iter_->key().str();
+                val = iter_->val().str();
+            } else {
+                (*recoverEdgesIter_)->withRLock([&](auto&& data){
+                    key = data.first;
+                    val = data.second;
+                });
+            }
+        }
+        return ret;
+    }
+
+    void next() override {
+        if (!moveToValidRecord_) {
+            // ToDo(liuyu): what if meet a lock at the first time
```

Comment:
may be **update** also need to scan all the locks before the first edge.

Reply:
Exactly, for update, we only `need to scan all the locks before the first edge`, but I suppose you can't set `lookupOne_ = false` before you met the first edge, otherwise, update will think no edge at all. @panda-sheep, would you have a look at this piece of code.